Fix install with mbstring

eXorus opened the pull request:
Hello,

Fix this issue
https://bugs.php.net/bug.php?id=71813

On a fresh install of Ubuntu 17.10
```
sudo apt install php7.1-cli php-pear php-dev php-mbstring
sudo pecl install mailparse
[sudo] Mot de passe de exorus : 
WARNING: channel "pecl.php.net" has updated its protocols, use "pecl channel-update pecl.php.net" to update
downloading mailparse-3.0.2.tgz ...
Starting to download mailparse-3.0.2.tgz (38,206 bytes)
..........done: 38,206 bytes
10 source files, building
running: phpize
Configuring for:
PHP Api Version:         20160303
Zend Module Api No:      20160303
Zend Extension Api No:   320160303
building in /tmp/pear/temp/pear-build-rooteSkhbA/mailparse-3.0.2
running: /tmp/pear/temp/mailparse/configure --with-php-config=/usr/bin/php-config
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for a sed that does not truncate output... /bin/sed
checking for cc... cc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether cc accepts -g... yes
checking for cc option to accept ISO C89... none needed
checking how to run the C preprocessor... cc -E
checking for icc... no
checking for suncc... no
checking whether cc understands -c and -o together... yes
checking for system library directory... lib
checking if compiler supports -R... no
checking if compiler supports -Wl,-rpath,... yes
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking target system type... x86_64-pc-linux-gnu
checking for PHP prefix... /usr
checking for PHP includes... -I/usr/include/php/20160303 -I/usr/include/php/20160303/main -I/usr/include/php/20160303/TSRM -I/usr/include/php/20160303/Zend -I/usr/include/php/20160303/ext -I/usr/include/php/20160303/ext/date/lib
checking for PHP extension directory... /usr/lib/php/20160303
checking for PHP installed headers prefix... /usr/include/php/20160303
checking if debug is enabled... no
checking if zts is enabled... no
checking for re2c... no
configure: WARNING: You will need re2c 0.13.4 or later if you want to regenerate PHP parsers.
checking for gawk... no
checking for nawk... nawk
checking if nawk is broken... no
checking whether to enable mailparse support... yes, shared
checking how to print strings... printf
checking for a sed that does not truncate output... (cached) /bin/sed
checking for fgrep... /bin/grep -F
checking for ld used by cc... /usr/bin/ld
checking if the linker (/usr/bin/ld) is GNU ld... yes
checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm -B
checking the name lister (/usr/bin/nm -B) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 1572864
checking how to convert x86_64-pc-linux-gnu file names to x86_64-pc-linux-gnu format... func_convert_file_noop
checking how to convert x86_64-pc-linux-gnu file names to toolchain format... func_convert_file_noop
checking for /usr/bin/ld option to reload object files... -r
checking for objdump... objdump
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for ar... ar
checking for archiver @FILE support... @
checking for strip... strip
checking for ranlib... ranlib
checking for gawk... (cached) nawk
checking command to parse /usr/bin/nm -B output from cc object... ok
checking for sysroot... no
checking for a working dd... /bin/dd
checking how to truncate binary pipes... /bin/dd bs=4096 count=1
checking for mt... mt
checking if mt is a manifest tool... no
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for dlfcn.h... yes
checking for objdir... .libs
checking if cc supports -fno-rtti -fno-exceptions... no
checking for cc option to produce PIC... -fPIC -DPIC
checking if cc PIC flag -fPIC -DPIC works... yes
checking if cc static flag -static works... yes
checking if cc supports -c -o file.o... yes
checking if cc supports -c -o file.o... (cached) yes
checking whether the cc linker (/usr/bin/ld -m elf_x86_64) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... no
configure: creating ./config.status
config.status: creating config.h
config.status: executing libtool commands
running: make
/bin/bash /tmp/pear/temp/pear-build-rooteSkhbA/mailparse-3.0.2/libtool --mode=compile cc  -I. -I/tmp/pear/temp/mailparse -DPHP_ATOM_INC -I/tmp/pear/temp/pear-build-rooteSkhbA/mailparse-3.0.2/include -I/tmp/pear/temp/pear-build-rooteSkhbA/mailparse-3.0.2/main -I/tmp/pear/temp/mailparse -I/usr/include/php/20160303 -I/usr/include/php/20160303/main -I/usr/include/php/20160303/TSRM -I/usr/include/php/20160303/Zend -I/usr/include/php/20160303/ext -I/usr/include/php/20160303/ext/date/lib  -DHAVE_CONFIG_H  -g -O2   -c /tmp/pear/temp/mailparse/mailparse.c -o mailparse.lo
libtool: compile:  cc -I. -I/tmp/pear/temp/mailparse -DPHP_ATOM_INC -I/tmp/pear/temp/pear-build-rooteSkhbA/mailparse-3.0.2/include -I/tmp/pear/temp/pear-build-rooteSkhbA/mailparse-3.0.2/main -I/tmp/pear/temp/mailparse -I/usr/include/php/20160303 -I/usr/include/php/20160303/main -I/usr/include/php/20160303/TSRM -I/usr/include/php/20160303/Zend -I/usr/include/php/20160303/ext -I/usr/include/php/20160303/ext/date/lib -DHAVE_CONFIG_H -g -O2 -c /tmp/pear/temp/mailparse/mailparse.c  -fPIC -DPIC -o .libs/mailparse.o
/tmp/pear/temp/mailparse/mailparse.c:34:2: error: #error The mailparse extension requires the mbstring extension!
 #error The mailparse extension requires the mbstring extension!
  ^~~~~
Makefile:200 : la recette pour la cible « mailparse.lo » a échouée
make: *** [mailparse.lo] Erreur 1
ERROR: `make' failed

```

eXorus committed:
Fix install with mbstring

eXorus commented:
@Sean-Der Do you still have time for this project ?

jstanden commented:
I'm not a maintainer, but you can do this in the meantime:
https://stackoverflow.com/questions/35793216/installing-mailparse-php7-mbstring-error/36636332#36636332

eXorus commented:
Thanks I already know this trick but it's difficult to explain to my user
that I'm using an extension that is doesn't work by default.

Are you able to merge my PR? Could I help ?

Le lun. 4 déc. 2017 à 23:35, Jeff Standen <notifications@github.com> a
écrit :

> I'm not a maintainer, but you can do this in the meantime:
>
> https://stackoverflow.com/questions/35793216/installing-mailparse-php7-mbstring-error/36636332#36636332
>
> —
> You are receiving this because you authored the thread.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/php/pecl-mail-mailparse/pull/4#issuecomment-349129715>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABMoiREZtoxyHqVyY6AwajbICFiifoDuks5s9HOngaJpZM4QOorD>
> .
>

jstanden commented:
@eXorus I'm not a mailparse developer, but my project also has 1000+ clients using this extension and running into the compile issue more often with the newer Ubuntu versions.

This project feels abandoned, but I do hope they merge your PR, since it would help us as well.

johannes commented:
I don'T know much about this module and don't know if the dependency is needed, but if this is mostly about load order or pure runtime dependency thn this probably should be marked using `ZEND_MOD_REQUIRED_EX` (see other PHP modules for examples on usage) if there is absolutely no need for mbstring this can indeed be removed.

eXorus commented:
@johannes the dependency mbstring is needed and libmbfl is necessary for mbstring.
http://php.net/manual/en/mbstring.installation.php

So the condition is good for me I did the test. Everyone have issue about the install of mailparse so it's better than nothing.

Sorry I don't find any usage of ZEND_MOD_REQUIRED_EX.

mstaack commented:
@wez @Sean-Der  any updates on this one? would love to have this fixed..most people fail to install mailparse because of this. thanks max

wez commented:
Errrr, I'm very stale on this.  mailparse needed mbstring for its transcoding functions.  I don't know what the right way to express that is for php extensions these days.  If the configure stuff is good enough its possible that that additional check in the code is no longer needed.

eXorus commented:
@johannes According to this http://php.theraven7.com/manual/fr/zend-macro.zend-mod-required.php there is no difference between ZEND_MOD_REQUIRED and ZEND_MOD_REQUIRED_EX.

According to this https://github.com/php/php-src/blob/b864e6b58c2e1a7d2679412ad832903015961d3b/ext/pdo_mysql/pdo_mysql.c or this https://github.com/mongodb/mongo-php-driver/blob/2be8af5d42a25af949851789f6b23cd7af67d4aa/php_phongo.c

Maybe the solution could be to have that in mailparse.c:
- Added new const zend_module_dep
- Changed STANDARD_MODULE_HEADER to STANDARD_MODULE_HEADER_EX
- Added mailparse_deps before mailparse

```
static const zend_module_dep mailparse_deps[] = {
	ZEND_MOD_REQUIRED("mbstring")
	ZEND_MOD_REQUIRED("json")
	ZEND_MOD_REQUIRED("spl")
	ZEND_MOD_REQUIRED("standard")
	ZEND_MOD_END
};

zend_module_entry mailparse_module_entry = {
	STANDARD_MODULE_HEADER_EX, NULL,
	mailparse_deps
	"mailparse",
	mailparse_functions,
	PHP_MINIT(mailparse),
	PHP_MSHUTDOWN(mailparse),
	PHP_RINIT(mailparse),
	PHP_RSHUTDOWN(mailparse),
	PHP_MINFO(mailparse),
	PHP_MAILPARSE_VERSION,
	STANDARD_MODULE_PROPERTIES
};

```

And with that it will be possible to remove the extra line that I added in this PR.
I found the solution on github https://github.com/search?p=9&q=ZEND_MOD_REQUIRED&type=Code&utf8=%E2%9C%93

I can make the change and test it but are you agree to merge it or not ? @wez @johannes @Sean-Der ?

eXorus commented:
🎆 Happy new year 🎆 
@wez @johannes @Sean-Der

nicolasbadia commented:
For people still looking for a solution, here is another option: https://github.com/zbateson/mail-mime-parser

eXorus commented:
@nicolasbadia yes very nice package but mailparse is faster x2
https://github.com/zbateson/mail-mime-parser/issues/87

spk commented:
Its still happening with v3.1.4 and updated Debian/Ubuntu using shared-lib mbstring, any interest if i resubmit the patch?

Look like `sed -i 's/#if\s!HAVE_MBSTRING/#ifndef MBFL_MBFILTER_H/' ./mailparse.c ` do the trick too
